worldedit-core/src/main/java/com/sk89q/worldedit/internal/schematic/backends/FileWatcherSchematicsBackend.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * WorldEdit, a Minecraft world manipulation toolkit
+ * Copyright (C) sk89q <http://www.sk89q.com>
+ * Copyright (C) WorldEdit team and contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package com.sk89q.worldedit.internal.schematic.backends;
+
+import com.sk89q.worldedit.internal.util.LogManagerCompat;
+import com.sk89q.worldedit.internal.util.RecursiveDirectoryWatcher;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * A backend that efficiently scans for file changes using {@link RecursiveDirectoryWatcher}.
+ */
+public class FileWatcherSchematicsBackend implements SchematicsBackend {
+    private static final Logger LOGGER = LogManagerCompat.getLogger();
+
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    private final Set<Path> schematics = new HashSet<>();
```

Comment:
Why are we not using a `Sets.newConcurrentHashSet()`? It would work much better since it doesn't need to lock the whole set.